certbot/certbot/interfaces.py

```diff
@@ -532,6 +532,62 @@ def print_messages(self):
         """Prints messages to the user and clears the message queue."""
 
 
+@six.add_metaclass(abc.ABCMeta)
+class RenewableCert(object):
+    """Interface to a certificate lineage."""
+
+    @abc.abstractproperty
```

Comment:
Just to be sure.

Before, the result of these properties could be modified by assigning new values to the internal fields that back them. Now, from the interface point of view, there is no constraint nor indication about the capability to modify them (no setter contract for instance).

It is fine for me, since having a mutable or immutable concrete class behind the interface is an implementation detail in my opinion, but do you confirm that is also what you want to express here?

Reply:
These properties should be treated as read only.

I think this is the normal behavior people expect from a Python class. How often do you create a class and start modifying attributes of that class without being explicitly told you can do so? I'd argue that the absence of being told you can modify these values and there only being read only properties defined loosely but reasonably expresses that you shouldn't modify these values.

I don't think there's really a (sane) way we can enforce this behavior in the concrete class as any subclass can provide their own setter methods overriding any in this class.

The best idea I have if you think this should be more explicitly expressed is to add "read only" to the docstring of each property, but I personally think this is overly verbose and largely just restates what we have in code about there only being read only properties defined.

Reply:
> I think this is the normal behavior people expect from a Python class. How often do you create a class and start modifying attributes of that class without being explicitly told you can do so? I'd argue that the absence of being told you can modify these values and there only being read only properties defined loosely but reasonably expresses that you shouldn't modify these values.

Right for a class, less for an interface, where mutability is not applying. I would never mess up with a class that is designed to be immutable. However here it is an interface, and the meaning is slightly different: the interface states the existence of getters, but does not say anything about the setters.

It would be perfectly valid to write a mutable class that implements the interface. It would be not for the logic here since Certbot is globally relying on the fact that these properties do not change after the object creation.

No, I do not think there is any sane way in any language to force a class to be immutable when it implements some interface.

However for Python specifically, where we are not talking about real interfaces, official documentation is strongly implying that using `@abc.abstractproperty` alone is associated to a read-only property (https://docs.python.org/3/library/abc.html#abc.abstractproperty). So I am fine with it ^^